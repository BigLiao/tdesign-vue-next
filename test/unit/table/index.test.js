import { mount } from '@vue/test-utils';
import Table from '@/src/table/index.ts';

const data = [
  {
    id: 1,
    type: 'Number',
    value: '1',
  },
  {
    id: 2,
    type: 'Number',
    value: '2',
  },
];
const columns = [
  {
    title: 'id',
    colKey: 'id',
  },
  {
    title: '类型',
    colKey: 'type',
  },
  {
    title: '取值',
    colKey: 'value',
  },
];
const pagination = {
  pageSize: 10,
  total: 120,
  visibleWithOnePage: true,
};
// every component needs four parts: props/events/slots/functions.
describe('Table', () => {
  // test props api
  describe(':props', () => {
    describe(':prop.data', () => {
      it('`data` is not undefined', () => {
        const wrapper = mount({
          render() {
            return <Table data={data} columns={columns}></Table>;
          },
        });
        expect(wrapper.vm.$el.getElementsByTagName('tr').length).toBe(data.length + 1);
      });
      it('`data` is undefined', () => {
        const empty = {
          customRender: 'empty',
        };
        const wrapper = mount({
          render() {
            return (
              <Table columns={columns} empty={empty}>
                <div slot="empty" id="empty-container">
                  暂无数据
                </div>
              </Table>
            );
          },
        });
        expect(wrapper.find('#empty-container').exists()).toBe(true);
      });
    });
    describe(':props.columns', () => {
      it('`columns` is undefined', () => {
        const wrapper = mount({
          render() {
            return <Table></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('`columns` with children', () => {
        const data = [
          {
            platform: '公有',
            property: 'data',
            type: 'any[]',
            default: '[]',
            needed: 'Y',
            description: '数据源',
          },
          {
            platform: '公有',
            property: 'rowkey',
            type: 'String',
            default: '-1',
            needed: 'Y',
            description: '指定rowkey',
          },
        ];
        const columns = [
          {
            title: '汇总属性',
            children: [
              {
                align: 'left',
                width: '100',
                minWidth: '100',
                className: 'test',
                ellipsis: true,
                colKey: 'platform',
                title: '平台',
              },
              {
                title: '类型及默认值',
                children: [
                  {
                    align: 'left',
                    width: '100',
                    minWidth: '100',
                    className: 'row',
                    ellipsis: true,
                    colKey: 'type',
                    title: '类型',
                  },
                  {
                    align: 'left',
                    width: '100',
                    minWidth: '100',
                    className: 'test4',
                    ellipsis: true,
                    colKey: 'default',
                    title: '默认值',
                  },
                  {
                    align: 'left',
                    width: '100',
                    minWidth: '100',
                    className: 'test3',
                    ellipsis: true,
                    colKey: 'needed',
                    title: '是否必传',
                  },
                ],
              },
            ],
          },
          {
            title: '属性及说明',
            children: [
              {
                align: 'left',
                width: '100',
                minWidth: '100',
                className: 'test2',
                ellipsis: true,
                colKey: 'property',
                title: '属性',
              },
              {
                align: 'left',
                width: '100',
                minWidth: '100',
                className: 'row',
                ellipsis: true,
                colKey: 'description',
                title: '说明',
              },
            ],
          },
        ];
        const wrapper = mount({
          render() {
            return <Table data={data} columns={columns}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('`columns` with fixed column', () => {
        const columns = [
          {
            title: 'id',
            colKey: 'id',
            fixed: 'left',
          },
          {
            title: '类型',
            colKey: 'type',
          },
          {
            title: '取值',
            colKey: 'value',
          },
        ];
        const wrapper = mount({
          render() {
            return <Table columns={columns} data={data}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('`columns` with align, width, minWidth and ellipsis', () => {
        const columns = [
          {
            title: 'id',
            colKey: 'id',
            width: 100,
            align: 'left',
            ellipsis: true,
            minWidth: 200,
          },
          {
            title: '类型',
            colKey: 'type',
          },
          {
            title: '取值',
            colKey: 'value',
          },
        ];
        const wrapper = mount({
          render() {
            return <Table columns={columns} data={data}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe(':props.pagination', () => {
      it('`pagination` visibleWithOnePage is true', () => {
        const wrapper = mount({
          render() {
            return <Table data={data} columns={columns} pagination={pagination}></Table>;
          },
        });
        expect(wrapper.find('.t-table-pagination').exists()).toBe(true);
      });
    });
    describe(':props.empty', () => {
      it('`empty` is undefined', () => {
        const wrapper = mount({
          render() {
            return <Table columns={columns}></Table>;
          },
        });
        expect(wrapper.find('.t-table--empty').exists()).toBe(true);
      });
      it('`empty` is a Function', () => {
        const empty = jest.fn(() => {
          /* */
        });
        mount({
          render() {
            return <Table columns={columns} empty={empty}></Table>;
          },
        });
        expect(empty.mock.calls.length).toBe(1);
      });
    });
    describe(':props.loading', () => {
      it('`loading` is synchronous', () => {
        const loading = true;
        const wrapper = mount({
          render() {
            return <Table columns={columns} loading={loading}></Table>;
          },
        });
        expect(wrapper.find('.t-table--loading').exists()).toBe(true);
      });
      it('`loading` is rendered with `slot`', () => {
        const wrapper = mount({
          render() {
            const loadingOpt = {
              customRender: 'loading',
            };
            const loading = () => <div class="loading__container">loading</div>;
            const scopedSlots = {
              loading,
            };
            return <Table loading={loadingOpt} scopedSlots={scopedSlots}></Table>;
          },
        });
        expect(wrapper.find('.loading__container').exists()).toBe(true);
      });
      it('`loading` is asynchronous', () => {
        const loading = {
          async: true,
        };
        const wrapper = mount({
          render() {
            return <Table columns={columns} loading={loading}></Table>;
          },
        });
        expect(wrapper.find('.t-table--loading-async').exists()).toBe(true);
      });
    });
    describe(':props.border, :props:hover, :props.stripe, :props.size, :props.verticalAlign, :props.height', () => {
      it('size = "small", verticalAlign="middle"', () => {
        const wrapper = mount({
          render() {
            return (
              <Table
                data={data}
                columns={columns}
                border={true}
                stripe={true}
                size={'small'}
                hover={true}
                verticalAlign={'middle'}
              ></Table>
            );
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('size = "large", verticalAlign="top"', () => {
        const wrapper = mount({
          render() {
            return <Table size={'large'} verticalAlign={'top'}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('size = "default", verticalAlign="bottom"', () => {
        const wrapper = mount({
          render() {
            return <Table size={'default'} verticalAlign={'bottom'}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
      it('height = 100', () => {
        const wrapper = mount({
          render() {
            return <Table height={100} data={data} columns={columns}></Table>;
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
  // test slots
  describe(':slots', () => {
    it('Use slot to customize cell', async () => {
      const wrapper = await mount({
        render() {
          const id = ({ text }) => <div class="cell">{text}</div>;
          const scopedSlots = { id };
          return <Table data={data} columns={columns} scopedSlots={scopedSlots}></Table>;
        },
      });
      expect(wrapper.findAll('.cell').length).toBe(data.length);
    });
    it('Use slot to customize header', async () => {
      const wrapper = await mount({
        render() {
          const slotColumns = columns.slice(0);
          slotColumns[0] = {
            colKey: 'id',
            slots: {
              title: 'columnId',
            },
          };
          const columnId = ({ text }) => <div class="header_cell">{text}</div>;
          const scopedSlots = { columnId };
          return <Table data={data} columns={slotColumns} scopedSlots={scopedSlots}></Table>;
        },
      });
      expect(wrapper.findAll('.header_cell').length).toBe(1);
    });
    it('Use slot to customize empty content', async () => {
      const wrapper = await mount({
        render() {
          const emptyProps = {
            customRender: 'empty',
          };
          const empty = () => <div class="empty__container">empty</div>;
          const scopedSlots = { empty };
          return <Table empty={emptyProps} columns={columns} scopedSlots={scopedSlots}></Table>;
        },
      });
      expect(wrapper.find('.empty__container').exists()).toBe(true);
    });
  });

  // sort
  describe(':sort', () => {
    const data = [
      {
        key: '1',
        name: 'John Brown',
        age: 32,
        address: 'New York No. 1 Lake Park',
      },
      {
        key: '2',
        name: 'Jim Green',
        age: 42,
        address: 'London No. 1 Lake Park',
      },
      {
        key: '3',
        name: 'Joe Black',
        age: 20,
        address: 'Sidney No. 1 Lake Park',
      },
    ];
    const columns = [
      {
        title: 'Name',
        colKey: 'name',
      },
      {
        title: 'Age',
        colKey: 'age',
        sorter: (a, b) => a.age - b.age,
      },
      {
        title: 'Address',
        colKey: 'address',
      },
    ];
    it('click sorter icon', async () => {
      const wrapper = await mount({
        render() {
          return <Table data={data} columns={columns} />;
        },
      });
      expect(wrapper.find('.table-body tr td').text()).toBe('John Brown');
      const clickDom = wrapper.find('thead tr td .t-table-sort-icon');
      await clickDom.trigger('click');
      expect(wrapper.find('.table-body tr td').text()).toBe('Joe Black');
      await clickDom.trigger('click');
      expect(wrapper.find('.table-body tr td').text()).toBe('Jim Green');
    });
  });

  // asyncLoading
  describe(':asyncLoading', () => {
    let wrapper;

    it('init status without async loading row', async () => {
      wrapper = await mount({
        data() {
          return {
            asyncLoading: false,
          };
        },
        render() {
          const { asyncLoading } = this;
          return <Table data={data} columns={columns} asyncLoading={asyncLoading} />;
        },
      });
      expect(wrapper.find('.table-body tr td .t-table--loading-async').exists()).toBe(false);
    });

    it('with async loading row', async () => {
      await wrapper.setData({ asyncLoading: true });
      expect(wrapper.find('.table-body tr td .t-table--loading-async').exists()).toBe(true);
    });
  });

  // filters
  describe(':filters', () => {
    let wrapper;
    const data = [
      {
        key: '1',
        name: 'John Brown',
        age: 32,
        address: 'New York No. 1 Lake Park',
      },
      {
        key: '2',
        name: 'Jim Green',
        age: 42,
        address: 'London No. 1 Lake Park',
      },
      {
        key: '3',
        name: 'Joe Black',
        age: 32,
        address: 'Sidney No. 1 Lake Park',
      },
      {
        key: '4',
        name: 'Jim Red',
        age: 32,
        address: 'London No. 2 Lake Park',
      },
    ];

    it('controlled filters', async () => {
      wrapper = await mount({
        data() {
          return {
            data,
            filteredInfo: null,
          };
        },
        computed: {
          columns() {
            let { filteredInfo } = this;
            filteredInfo = filteredInfo || {};
            const columns = [
              {
                title: 'Name',
                colKey: 'name',
                key: 'name',
                filters: [
                  { label: 'Joe', value: 'Joe' },
                  { label: 'Jim', value: 'Jim' },
                ],
                filteredValue: filteredInfo.name || null,
                onFilter: (value, record) => record.name.includes(value),
                filterMultiple: true,
              },
              {
                title: 'Age',
                colKey: 'age',
                key: 'age',
              },
              {
                title: 'Address',
                colKey: 'address',
                key: 'address',
                filters: [
                  { label: 'London', value: 'London' },
                  { label: 'New York', value: 'New York' },
                ],
                filteredValue: filteredInfo.address || null,
                onFilter: (value, record) => record.address.includes(value),
              },
            ];
            return columns;
          },
        },
        render() {
          const { data, columns } = this;
          return <Table data={data} columns={columns} />;
        },
      });
      expect(wrapper.findAll('.table-body tr').length).toBe(data.length);
      await wrapper.setData({ filteredInfo: { name: ['Jim'] } });
      expect(wrapper.findAll('.table-body tr').length).toBe(data.filter(({ name }) => name.includes('Jim')).length);
      await wrapper.setData({ filteredInfo: null });
      expect(wrapper.findAll('.table-body tr').length).toBe(data.length);
    });
  });
});
