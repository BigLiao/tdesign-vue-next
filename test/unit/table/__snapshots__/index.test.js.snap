// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table :props :props.border, :props:hover, :props.stripe, :props.size, :props.verticalAlign, :props.height height = 100 1`] = `
<div>
  <div class="t-table t-table__header--fixed">
    <div class="t-table-content">
      <div class="t-table__header">
        <table style="width: 100%;">
          <colgroup>
            <col>
            <col>
            <col>
          </colgroup>
          <thead columnsprops="[object Object],[object Object],[object Object]">
            <tr>
              <td style="" class="" key="id" colspan="1" rowspan="1">id</td>
              <td style="" class="" key="type" colspan="1" rowspan="1">类型</td>
              <td style="" class="" key="value" colspan="1" rowspan="1">取值</td>
            </tr>
          </thead>
        </table>
      </div>
      <div class="t-table__body" style="max-height: 100px;">
        <table style="width: 100%;">
          <colgroup>
            <col>
            <col>
            <col>
          </colgroup>
          <tbody class="table-body">
            <tr class="">
              <td style="" class="" key="id" length="3">1</td>
              <td style="" class="" key="type" length="3">Number</td>
              <td style="" class="" key="value" length="3">1</td>
            </tr>
            <tr class="">
              <td style="" class="" key="id" length="3">2</td>
              <td style="" class="" key="type" length="3">Number</td>
              <td style="" class="" key="value" length="3">2</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table :props :props.border, :props:hover, :props.stripe, :props.size, :props.verticalAlign, :props.height size = "default", verticalAlign="bottom" 1`] = `
<div>
  <div class="t-table t-table-valign__bottom">
    <div class="t-table-content">
      <table>
        <colgroup></colgroup>
        <thead columnsprops="">
          <tr></tr>
        </thead>
      </table>
    </div><b>暂无数据</b>
  </div>
</div>
`;

exports[`Table :props :props.border, :props:hover, :props.stripe, :props.size, :props.verticalAlign, :props.height size = "large", verticalAlign="top" 1`] = `
<div>
  <div class="t-table t-size-l t-table-valign__top">
    <div class="t-table-content">
      <table>
        <colgroup></colgroup>
        <thead columnsprops="">
          <tr></tr>
        </thead>
      </table>
    </div><b>暂无数据</b>
  </div>
</div>
`;

exports[`Table :props :props.border, :props:hover, :props.stripe, :props.size, :props.verticalAlign, :props.height size = "small", verticalAlign="middle" 1`] = `
<div>
  <div class="t-table t-table--bordered t-table--striped t-table--hoverable t-size-s">
    <div class="t-table-content">
      <table>
        <colgroup>
          <col>
          <col>
          <col>
        </colgroup>
        <thead columnsprops="[object Object],[object Object],[object Object]">
          <tr>
            <td style="" class="" key="id" colspan="1" rowspan="1">id</td>
            <td style="" class="" key="type" colspan="1" rowspan="1">类型</td>
            <td style="" class="" key="value" colspan="1" rowspan="1">取值</td>
          </tr>
        </thead>
        <tbody class="table-body">
          <tr class="">
            <td style="" class="" key="id" length="3">1</td>
            <td style="" class="" key="type" length="3">Number</td>
            <td style="" class="" key="value" length="3">1</td>
          </tr>
          <tr class="">
            <td style="" class="" key="id" length="3">2</td>
            <td style="" class="" key="type" length="3">Number</td>
            <td style="" class="" key="value" length="3">2</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Table :props :props.columns \`columns\` is undefined 1`] = `
<div>
  <div class="t-table">
    <div class="t-table-content">
      <table>
        <colgroup></colgroup>
        <thead columnsprops="">
          <tr></tr>
        </thead>
      </table>
    </div><b>暂无数据</b>
  </div>
</div>
`;

exports[`Table :props :props.columns \`columns\` with align, width, minWidth and ellipsis 1`] = `
<div>
  <div class="t-table">
    <div class="t-table-content">
      <table>
        <colgroup>
          <col style="text-overflow: ellipsis; display: -webkit-box; overflow: hidden; text-align: left; width: 100px; min-width: 200px;;">
          <col>
          <col>
        </colgroup>
        <thead columnsprops="[object Object],[object Object],[object Object]">
          <tr>
            <td style="text-align: left;" class="" key="id" colspan="1" rowspan="1">id</td>
            <td style="" class="" key="type" colspan="1" rowspan="1">类型</td>
            <td style="" class="" key="value" colspan="1" rowspan="1">取值</td>
          </tr>
        </thead>
        <tbody class="table-body">
          <tr class="">
            <td style="text-align: left;" class="" key="id" length="3">1</td>
            <td style="" class="" key="type" length="3">Number</td>
            <td style="" class="" key="value" length="3">1</td>
          </tr>
          <tr class="">
            <td style="text-align: left;" class="" key="id" length="3">2</td>
            <td style="" class="" key="type" length="3">Number</td>
            <td style="" class="" key="value" length="3">2</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Table :props :props.columns \`columns\` with children 1`] = `
<div>
  <div class="t-table">
    <div class="t-table-content">
      <table>
        <colgroup>
          <col>
          <col>
        </colgroup>
        <thead columnsprops="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]">
          <tr>
            <td style="" class="" colspan="4" rowspan="1">属性及说明</td>
            <td style="" class="" colspan="2" rowspan="1">属性及说明</td>
          </tr>
          <tr>
            <td style="text-align: left;" class="" key="platform" colspan="1" rowspan="2">平台</td>
            <td style="" class="" colspan="3" rowspan="1">类型及默认值</td>
            <td style="text-align: left;" class="" key="property" colspan="1" rowspan="2">属性</td>
            <td style="text-align: left;" class="" key="description" colspan="1" rowspan="2">说明</td>
          </tr>
          <tr>
            <td style="text-align: left;" class="" key="type" colspan="1" rowspan="1">类型</td>
            <td style="text-align: left;" class="" key="default" colspan="1" rowspan="1">默认值</td>
            <td style="text-align: left;" class="" key="needed" colspan="1" rowspan="1">是否必传</td>
          </tr>
        </thead>
        <tbody class="table-body">
          <tr class="">
            <td style="text-align: left;" class="" key="platform" length="6">公有</td>
            <td style="text-align: left;" class="" key="type" length="6">any[]</td>
            <td style="text-align: left;" class="" key="default" length="6">[]</td>
            <td style="text-align: left;" class="" key="needed" length="6">Y</td>
            <td style="text-align: left;" class="" key="property" length="6">data</td>
            <td style="text-align: left;" class="" key="description" length="6">数据源</td>
          </tr>
          <tr class="">
            <td style="text-align: left;" class="" key="platform" length="6">公有</td>
            <td style="text-align: left;" class="" key="type" length="6">String</td>
            <td style="text-align: left;" class="" key="default" length="6">-1</td>
            <td style="text-align: left;" class="" key="needed" length="6">Y</td>
            <td style="text-align: left;" class="" key="property" length="6">rowkey</td>
            <td style="text-align: left;" class="" key="description" length="6">指定rowkey</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Table :props :props.columns \`columns\` with fixed column 1`] = `
<div>
  <div class="t-table t-table__cell--fixed t-table--has-fixed">
    <div class="t-table-content">
      <div class="t-table__header">
        <table style="width: 100%;">
          <colgroup>
            <col>
            <col>
            <col>
          </colgroup>
          <thead columnsprops="[object Object],[object Object],[object Object]">
            <tr>
              <td style="position: sticky;left: 0px;" class="t-table__cell--fixed-left t-table__cell--fixed-left-last" key="id" colspan="1" rowspan="1">id</td>
              <td style="" class="" key="type" colspan="1" rowspan="1">类型</td>
              <td style="" class="" key="value" colspan="1" rowspan="1">取值</td>
            </tr>
          </thead>
        </table>
      </div>
      <div class="t-table__body" style="max-height: auto;">
        <table style="width: 100%;">
          <colgroup>
            <col>
            <col>
            <col>
          </colgroup>
          <tbody class="table-body">
            <tr class="">
              <td style="position: sticky;left: 0px;" class="t-table__cell--fixed-left t-table__cell--fixed-left-last" key="id" length="3">1</td>
              <td style="" class="" key="type" length="3">Number</td>
              <td style="" class="" key="value" length="3">1</td>
            </tr>
            <tr class="">
              <td style="position: sticky;left: 0px;" class="t-table__cell--fixed-left t-table__cell--fixed-left-last" key="id" length="3">2</td>
              <td style="" class="" key="type" length="3">Number</td>
              <td style="" class="" key="value" length="3">2</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
